### ISSUE TITLE
Problem with function types, <T>() => T extends X ? () : ()

### ISSUE THREAD

**Issue by gs18004** [microsoft/TypeScript#56721]
### 🔎 Search Terms

"function", "generic function"

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAYgdgHgBoD4oF4oICooBQCUGa2UEAHsBHACYDOUSUA-FAIxQBcUATAFChIUAKIBHAK4BDADZ02yADRQAmmnR8osRKjKVq9DVoSqWUYACdx0bgDMZdCHwHhoAdXMB7OAHMAShDpxaWAMEQl7eQBvcwhJGi9pEChJbgByAEFUgF8lSJSoDOyUJydBaHh5HUwcfCJ0El0qWgYmVg5ufjKtHmQ1LFxCYihSCib6RlN23mchMSlZHqQlE3VNCt7G-TpDeB6TVgsrLig7WUcZ6ABhD3MYgGNgf0Dg0Ln7HujY+LhE5LTMnJQPL-IpAA

### 💻 Code

```ts
type Fn<X> = <T>() => T extends X ? 1 : 2
type Equals<X, Y> =
  Fn<X> extends
  Fn<Y> ? true : false

type A = Equals<{readonly a: 'A'}, {a: 'A'}>
```

### 🙁 Actual behavior

type A should shows true.

### 🙂 Expected behavior

type A should be false.

It works well when the code is like this.
```ts
type Fn1<X> = <T>() => T extends X ? 1 : 2
type Fn2<X> = <T>() => T extends X ? 1 : 2
type Equals<X, Y> =
  Fn1<X> extends
  Fn2<Y> ? true : false

type A = Equals<{readonly a: 'A'}, {a: 'A'}>
```

**Comment by RyanCavanaugh** [comment]
Where are people getting this definition of `Equals` from??

**Comment by jcalz** [comment]
Probably originally from https://github.com/microsoft/TypeScript/issues/27024#issuecomment-421529650 and then from [sindresorhus/type-fest](https://github.com/sindresorhus/type-fest/blob/a919f93384dba6e6d21e202a287986a6d4ea03e4/source/is-equal.d.ts) after that?

**Comment by fatcerberus** [comment]
Context: This variant allows for distinguishing `any` from other types, which the more traditional variants of `Equals` don’t.

**Comment by mintway0341** [comment]
> Where are people getting this definition of `Equals` from??

I got it from https://github.com/type-challenges/type-challenges/blob/main/utils/index.d.ts

**Comment by craigphicks** [comment]
@fatcerebus
> Context: This variant allows for distinguishing `any` from other types, which the more traditional variants of `Equals` don’t.

What is amounts to is that the "correct" answer(s) to a type function "Equals(A,B)" depends on what it is being used for.

In particular, the types `any`, `never`, `unknown`, and types described by an intersection `&` operator are corner cases that make Equals hard to define as a one-size-fits-all.

Therefore exporting these intrinsic type functions
- `isAnyType`
- `isNeverType`
- `isUnknownType`
- `isIntersectionType`
- `getIntersectionTypes`
 
would enable the users to handle the corner cases as they need, and simply apply `extends` to the remainder.

**Comment by fatcerberus** [comment]
@craigphicks That’s all true, but TS doesn’t add new pack-in utility types (intrinsic or otherwise) unless needed for declaration emit, so we have to work with what we have.

**Comment by craigphicks** [comment]
@fatcerberus - Maybe that should have been a sad emoji.

**Comment by craigphicks** [comment]
I am looking through the documentation for a good definition of `extends` (as used in conditional expressions) and I can't find it.

*Guess1*: One guess is that it is subset applied to the set of readable values, in which case `{readonly a: 'A'}` and  `{a: 'A'}` are mutually extendable.  


*Guess2* : Another guess is that it is subset applied to both the set of readable values and inverse-subset of writable values:
```
L extends R ::= readableValues(L) subsetOf readableValues(R) && writableValues(R) subsetOf writableValues(L)
```

In fact
```
type T = Readonly<{readonly a: 'A'}> extends {a:'A'} ? true : false; // true
```
evaluates to `true`, so that doesn't match Guess2.  

*Guess3*:  Like Guess2, but to detect the readability of the inner properties it is necessary to do it indirectly using an intermediate function like `Fn<X> = <T>() => T extends X`.  

There ought to be an easier way, not only for the sake of the user, but also for the sake of the implementer.

**Comment by craigphicks** [comment]
> @craigphicks That’s all true, but TS doesn’t add new pack-in utility types (intrinsic or otherwise) unless needed for declaration emit, so we have to work with what we have.

If that because of a with not to pollute the global namespace?  That's legit, but using a namespace (e.g. `ts.`) to prefix all utilities would solve it.

**Comment by RyanCavanaugh** [comment]
The reason for not adding utility types is that whatever definition we choose will invariably not match what 30% (minimum) of people expect to happen w.r.t how they treat any, whether they distribute over unions, how they treat `unknown`/`void`, whether the mapped forms are homomorphic, etc, leading to unending complaints that we picked the "wrong" definition of any given utility type.

**Comment by RyanCavanaugh** [comment]
It seems like we're using the strict subtype relation in one case, and regular assignability in the other. I don't even know which people are wanting with this definition of `Equals`! It's perhaps possible to unify to the same relation but it's very difficult to predict what would break in that case.

I'd really encourage everyone to just use something like
```ts
type Equals<X, Y> = [X] extends [Y] ? [Y] extends [X] ? true: false;
```
if they're not actively trying to probe the strangest corners of the type system 😵‍💫

**Comment by fatcerberus** [comment]
> I don't even know which [type relation] people are wanting with this definition of Equals

The intent of this definition of `Equals` is ultimately to do `isTypeIdenticalTo(X, Y)`; I believe the correct path for that is assignability (specifically, of deferred conditional types)
